libvast/src/system/query_supervisor.cpp

```diff
@@ -48,33 +48,41 @@ query_supervisor_actor::behavior_type query_supervisor(
   // Ask master for initial work.
   self->send(master, atom::worker_v, self);
   return {
-    [=](const expression&, const query_map& qm,
+    [=](const expression& expr,
+        const std::vector<std::pair<uuid, partition_actor>>& qm,
         const index_client_actor& client) {
       VAST_DEBUG(self, "got a new query for", qm.size(),
                  "partitions:", get_ids(qm));
       VAST_ASSERT(!qm.empty());
-      VAST_ASSERT(self->state.open_requests.empty());
-      for (auto& kvp : qm) {
-        auto& id = kvp.first;
-        auto& evaluators = kvp.second;
-        VAST_DEBUG(self, "asks", evaluators.size(),
-                   "EVALUATOR actor(s) for partition", id);
-        self->state.open_requests.emplace(id, evaluators.size());
-        for (auto& evaluator : evaluators)
-          self->request(evaluator, caf::infinite, client).then([=](atom::done) {
-            auto& num_evaluators = self->state.open_requests[id];
-            if (--num_evaluators == 0) {
-              VAST_DEBUG(self, "collected all results for partition", id);
-              self->state.open_requests.erase(id);
-              // Ask master for more work after receiving the last sub
-              // result.
-              if (self->state.open_requests.empty()) {
-                VAST_DEBUG(self, "collected all results for all partitions");
+      VAST_ASSERT(self->state.open_requests == 0);
+      for (auto& [id, partition] : qm) {
+        ++self->state.open_requests;
+        // TODO: Add a proper configurable timeout.
+        self
+          ->request(partition, caf::infinite, expr,
+                    caf::actor_cast<partition_client_actor>(client))
```

Comment:
```suggestion
                    static_cast<partition_client_actor>(client))
```

Reply:
Why?

Reply:
Because it suffices. `actor_cast` can cast to typed actors whose behaviors aren't subsets, while `static_cast` only works if the target behavior is a subset of the source behavior.

Reply:
Good to know.